### DISCUSSION TITLE
v1.2 Release Scoping: Features Graduating to Standard

### DISCUSSION THREAD

**Discussion by robscott** [kubernetes-sigs/gateway-api#3128] [General]
This discussion is similar to https://github.com/kubernetes-sigs/gateway-api/discussions/3103, but somewhat simpler. We'll track every GEP that could potentially graduate to standard in this release, and then look for owners to volunteer to drive them forward. Each top level comment will describe a GEP + what is required to take it to standard. To volunteer as a lead for that GEP, please add a comment in the thread below the GEP description.

As a reminder, each GEP that graduates to standard creates a new slot for a GEP to graduate from experimental (#3103), so if you really want a new experimental feature, volunteering to lead an experimental feature to standard will help. 

Note that inclusion in the list below is not a guarantee that a GEP will graduate in v1.2, just an acknowledgement that it theoretically could with sufficient effort and consensus.

**Comment by robscott** [comment]
# Timeouts (https://github.com/kubernetes-sigs/gateway-api/issues/1742) and Duration format (https://github.com/kubernetes-sigs/gateway-api/issues/2257).

**Status:**
Conformance: Complete
Documentation: Complete
Implementations: ?/3

**Work Required:** 
* Survey implementations to ensure we have enough (3+) and/or encourage more implementations to support this feature. 
* Updates to GEP, docs, and API types to bump status to standard.

**Previous Leads:**
@kflynn and @frankbu helped get these to experimental.

**Reply by kflynn** [reply]
I can take these two. 🙂

**Comment by robscott** [comment]
# GatewayClass Supported Features (https://github.com/kubernetes-sigs/gateway-api/issues/2162)

**Status:**
Conformance: N/A
Documentation: Not Started
Implementations: ?/3

**Work Required:** 
* Survey implementations to ensure we have enough (3+) and/or encourage more implementations to support this feature. 
* Determine if we need more structured format here such as grouping by resource and status (for example HTTPRoute.Experimental, HTTPRoute.Extended)
* Write documentation
* Updates to GEP and API types to bump status to standard.

**Previous Leads:**
@LiorLieberman has led this effort.

**Reply by kflynn** [reply]
Do we really _need_ to answer the structured-format question to get this over the line? Presumably we could go out with the flat format and then change if we later need to...

**Reply by dprotaso** [reply]
> Presumably we could go out with the flat format and then change if we later need to...

I don't have a preference but you could change it now to support future extensibility by changing it from a list of strings to a list of maps


eg. 
```yaml
supportedFeatures:
- HTTPRouteHostRewrite
- HTTPRouteMethodMatching
```

Becomes
```yaml
supportedFeatures:
- name: HTTPRouteHostRewrite
- name: HTTPRouteMethodMatching
```

**Reply by LiorLieberman** [reply]
I agree that just a list is something we can and should improve but I think I am plus 1 to @kflynn's comment (as long as we can change the status api later)

I feel like we are doing nothing if we change it now without concrete feedback. 

@arkodg just posted his thoughts [here](https://github.com/kubernetes-sigs/gateway-api/issues/2162#issuecomment-2148584241)

Regarding implementations, we have istio and envoy-gw. I also talked with @mikemorris and @snehachhabria about Azure Application Gateway. But yes, we still need one more implementation.

Also, as discussed, you can add to the work required - to move the conformance tests to read from GWC so more implementations would want to implement that.

**Reply by LiorLieberman** [reply]
Scoping is by June 11th, when we aiming to release ?

**Reply by robscott** [reply]
if we pick the middle of each of the ranges in https://gateway-api.sigs.k8s.io/contributing/release-cycle/#2-gep-iteration-and-review, that would be 6 + 4 + 3 weeks, so 13 weeks from June 11, which is apparently September 10. Too early to commit to that specific date, but I think September is the goal

**Reply by LiorLieberman** [reply]
opened a dedicated issue for this #3164

**Comment by robscott** [comment]
# BackendTLSPolicy (https://github.com/kubernetes-sigs/gateway-api/issues/1897)

**Status:**
Conformance: Not Started
Documentation: Mostly Complete
Implementations: 2/3

**Work Required:** 
* Survey implementations to ensure we have enough (3+) and/or encourage more implementations to support this feature. 
* Write conformance tests
* Updates to GEP, docs, and API types to bump status to standard.

**Previous Leads:**
@candita has led this effort.

**Reply by candita** [reply]
Found out we have a 3rd implementation already (NGINX, Contour, Envoy Gateway).  I will lead this for a v1.2 release.

**Reply by arkodg** [reply]
+1

**Reply by whitneygriffith** [reply]
@candita do you need help with the conformance tests? I'd love to work on it.

**Reply by whitneygriffith** [reply]
@robscott @youngnick does the #3138 plan for conformance testing suffice for this GEP to be merged as implementable for Standard?

**Reply by youngnick** [reply]
The plan seems pretty reasonable to me, probably want some thoughts from @mlavacca when he's back as well.

**Reply by mlavacca** [reply]
I left a couple of comments in https://github.com/kubernetes-sigs/gateway-api/issues/3138. I think the list is not only pretty complete, but some tests have to be removed from there (details in the issue).

**Reply by candita** [reply]
This did not make it to v1.2 Standard.

**Comment by robscott** [comment]
# Backend Protocol Selection (https://github.com/kubernetes-sigs/gateway-api/issues/1911)

**Status:**
Conformance: In Progress
Documentation: In Progress
Implementations: ?/3

**Work Required:** 
* Survey implementations to ensure we have enough (3+) and/or encourage more implementations to support this feature. 
* Complete documentation and conformance (this may already be done)
* Updates to GEP, docs, and API types to bump status to standard.

**Previous Leads:**
@dprotaso has led this effort.

**Reply by dprotaso** [reply]
Supported implementations (passing h2c & websocket test)
- envoy gateway
- istio
- contour

**Comment by robscott** [comment]
# TLSRoute (https://github.com/kubernetes-sigs/gateway-api/issues/2643)

**Status:**
Conformance: In Progress
Documentation: In Progress
Implementations: 4+/3

**Work Required:** 
* Complete documentation and conformance
* TLSRoute predates GEPs, but it would be useful to have a basic one for this + to cover why we don't want to just reuse TCPRoute for passthrough
* Updates to GEP, docs, and API types to bump status to standard.

**Previous Leads:**
It's been many years, I'm not really sure at this point...

**Reply by mlavacca** [reply]
I synced with @candita, who is the assignee of #2643, as I want to participate in moving it forward. We discussed a possible next step in the test suite which is to go through the list of expected conformance tests in https://github.com/kubernetes-sigs/gateway-api/issues/1579, check its comprehensiveness, and in case update it.

**Reply by arkodg** [reply]
+1

**Reply by mlavacca** [reply]
We discussed this in the meeting on June 4th, and the outcome is that I will lead this one

**Reply by ahmadalkhansa** [reply]
Hello, is tlsroute implemented in v1.3.0? If yes, I would eagerly install the related crd.

**Reply by mlavacca** [reply]
The plan is to graduate TLSRoute for 1.2. As a result, it will move from the experimental channel to the standard one. Can I ask you which implementation you are using/planning to use?

**Reply by ahmadalkhansa** [reply]
My apologies. I've been following the documentation of nginx-gateway-fabric v1.3.0 and it seems that their kustomization of the standard channel is pointing to the crd (https://github.com/kubernetes-sigs/gateway-api/tree/v1.1.0/config/crd/standard) of the standard channel. I think that I have to switch to the experimental channel. Extreme apologies for the confusion. Thank you and much appreciation for the great work that you are doing!!

**Reply by mlavacca** [reply]
Yeah no worries at all, @ahmadalkhansa! For the time being, if you want to use `TLSRoute`s you need to install the experimental channel of the CRDs. In case we'll manage to promote such an API for 1.2, they'll be available in the standard channel.

**Reply by candita** [reply]
This did not make it into release v1.2

**Comment by robscott** [comment]
# In-Cluster Gateway Config (https://github.com/kubernetes-sigs/gateway-api/issues/1762)

**Status:**
Conformance: N/A
Documentation: Not Started
Implementations: ?/3

**Work Required:** 
* Survey implementations to ensure we have enough (3+) and/or encourage more implementations to support this feature.
* Write documentation
* Updates to GEP, docs, and API types to bump status to standard.

**Previous Leads:**
@howardjohn led this to experimental status.

**Reply by keithmattix** [reply]
/cc @whitneygriffith 

I'll own this one. I think we also need conformance tests

**Reply by robscott** [reply]
I think it's ~impossible to write conformance tests for this, but @dprotaso did try with https://github.com/kubernetes-sigs/gateway-api/pull/2845. Worth revisiting if we missed anything in that discussion though.

**Reply by keithmattix** [reply]
Oh wow I wasn't familiar with that context...yeah this one is tricky. If we're all fine with no conformance tests for this one, that works for me too

**Reply by mikemorris** [reply]
To be clear, is the scope for this just the `spec.infrastructure` stanza with `annotations` and `labels` keys?

**Reply by keithmattix** [reply]
Yep, that's my understanding

**Reply by mikemorris** [reply]
Okay I think the issue link for this isn't quite correct then and this should be tracking [the no-longer-quite-accurately-named] https://github.com/kubernetes-sigs/gateway-api/issues/1762 and [GEP-1762](https://gateway-api.sigs.k8s.io/geps/gep-1762/) for those fields (which also includes some `spec.addresses` behavior with regard to an underlying Service resource and setting a well-known label on in-cluster resources generated by from a Gateway, which could actually help with "generated resources" flag mentioned in the [meeting notes](https://docs.google.com/document/d/1eg-YjOHaQ7UD28htdNxBR3zufebozXKyI28cl2E11tU/edit#heading=h.f0iw4a33ls12) - just check for matching annotations and labels on all - zero or more - resources with a matching label).

https://github.com/kubernetes-sigs/gateway-api/issues/1867 and [GEP-1867](https://gateway-api.sigs.k8s.io/geps/gep-1867/) track the prerequisite `spec.infrastructure` field (which we _could_ roll in if we only want to target GEP-1762 with this effort) but importantly mostly focuses on adding `spec.infrastructure.parametersRef` to Gateway objects, which I'm unsure if we're ready to promote yet (would need to check implementations separately).

**Reply by robscott** [reply]
Good catch, thanks @mikemorris! I've updated the original comment to point to #1762.
